pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java

```diff
@@ -44,15 +46,105 @@ public class ConsistentHashingStickyKeyConsumerSelector implements StickyKeyCons
         this.hashRing = new TreeMap<>();
+        this.consumerSelectionCounters = new WeakHashMap<>();
         this.numberOfPoints = numberOfPoints;
     }
 
+    /**
+     * This class is used to store the consumers and the selected consumer for a hash value in the hash ring.
+     * This attempts to distribute the consumers evenly in the hash ring for consumers with the same
+     * consumer name and priority level. These entries collide in the hash ring.
+     * The selected consumer is the consumer that is selected to serve the hash value.
+     * It is not changed unless a consumer is removed or a colliding consumer with higher priority or
+     * lower selection count is added.
+     */
+    private static class HashRingEntry {
+        // This class is used to store the consumer added to the hash ring
+        // sorting will be by priority, consumer name and active "selection" count of the consumer instance
+        // so that consumers get evenly distributed
+        record ConsumerEntry(Consumer consumer, MutableInt consumerSelectionCounter)
+                implements Comparable<ConsumerEntry> {
+            private static final Comparator<ConsumerEntry> CONSUMER_ENTRY_COMPARATOR =
+                    Comparator.<ConsumerEntry, Integer>
+                            // priority level is the primary sorting key
+                                    comparing(entry -> entry.consumer().getPriorityLevel()).reversed()
+                            // consumer name is the secondary sorting key
+                            .thenComparing(entry -> entry.consumer().consumerName())
+                            // then prefer the consumer instance with lowest selection count
+                            // so that consumers get evenly distributed
+                            .thenComparing(ConsumerEntry::consumerSelectionCounter);
+
+            @Override
+            public int compareTo(ConsumerEntry o) {
+                    return CONSUMER_ENTRY_COMPARATOR.compare(this, o);
+            }
+        }
+
+        private final List<ConsumerEntry> consumers;
+        ConsumerEntry selectedConsumerEntry;
+
+        public HashRingEntry() {
+            this.consumers = new ArrayList<>();
+        }
+
+        public Consumer getSelectedConsumer() {
+            return selectedConsumerEntry != null ? selectedConsumerEntry.consumer() : null;
+        }
+
+        public void addConsumer(Consumer consumer, MutableInt selectedCounter) {
+            ConsumerEntry consumerEntry = new ConsumerEntry(consumer, selectedCounter);
+            consumers.add(consumerEntry);
+            if (selectedConsumerEntry == null || consumerEntry.compareTo(selectedConsumerEntry) < 0) {
+                // if the new consumer has a higher priority or lower selection count
+                // than the currently selected consumer, select the new consumer
+                changeSelectedConsumerEntry(consumerEntry);
+            }
```

Comment:
I don't think that it's a breaking change. The previous behavior is already broken.

To fix the problem there needs to be a solution where the consumer is only changed when a new consumer is added or when the currently "selected" consumer is removed.

In the previous solution, all consumers for a colliding hash were active at the same time. The consumer was selected by the input hash on this line:
https://github.com/apache/pulsar/blob/4f96146f13b136644a4eb0cf4ec36699e0431929/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L127
That was a wrong solution that was broken.

One small detail what I made in the sorting order was that I made it also consider the priority of the consumer. That's been ignored before, but I think it has been a mistake before. Since the previous behavior has been broken, we aren't introducing a breaking change when we fix it.

Reply:
After your comment, I debugged your code, and I noticed that my understanding was incorrect.
My understanding at the time was, "If a new consumer is added, the range will always move to a new one."

Now, I understand as follows. Is it correct?
"When the selected count was greater than new consumer, then the consumer's range is moved to a new one."

Reply:
> After your comment, I debugged your code, and I noticed that my understanding was incorrect. My understanding at the time was, "If a new consumer is added, the range will always move to a new one."
> 
> Now, I understand as follows. Is it correct? "When the selected count was greater than new consumer, then the consumer's range is moved to a new one."

@equanz  In the previous solution `consumerList.get(hash % consumerList.size())` could result in a different consumer any time consumers are added or removed. Since the `hash` passed as the parameter is the hash calculated for the message key, it also depends on the input message. It could also result in the same consumer as before, so it doesn't mean that is always switched, however the probability increases when there are more consumers since hitting a different consumer has (n-1)/n chance.